packages/muelu/src/Transfers/SemiCoarsen/MueLu_SemiCoarsenPFactory_kokkos_decl.hpp

```diff
@@ -0,0 +1,154 @@
+  | semicoarsen: coarsen rate | int | * | * |   | Coarsening rate along vertical lines (2 corresponds to classical semicoarsening. Values > 2 for more aggressive coarsening). |
+
+  The * in the @c master.xml column denotes that the parameter is defined in the @c master.xml file.<br>
+  The * in the @c validated column means that the parameter is declared in the list of valid input parameters (see SemiCoarsenPFactory_kokkos::GetValidParameters).<br>
+  The * in the @c requested column states that the data is requested as input with all dependencies (see SemiCoarsenPFactory_kokkos::DeclareInput).
+
+  ### Variables provided by SemiCoarsenPFactory_kokkos ###
+  After SemiCoarsenPFactory_kokkos::Build the following data is available (if requested)
+
+  Parameter | generated by | description
+  ----------|--------------|------------
+  | P       | SemiCoarsenPFactory_kokkos   | Prolongator
+  | Nullspace | SemiCoarsenPFactory_kokkos | Coarse nullspace (the fine level nullspace information is coarsened using P to generate a coarse version of the nullspace. No scaling is applied.
+  | NumZLayers | NoFactory | Number of z layers after coarsening. Necessary input for LineDetectionFactory. Useful input for TogglePFactory.
+*/
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  class SemiCoarsenPFactory_kokkos;
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class DeviceType>
+  class SemiCoarsenPFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType>> : public PFactory {
+  public:
+    typedef LocalOrdinal                                        local_ordinal_type;
+    typedef GlobalOrdinal                                       global_ordinal_type;
+    typedef typename DeviceType::execution_space                execution_space;
+    typedef Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> node_type;
+
+  private:
+    // For compatibility
+    typedef node_type                                           Node;
+#undef MUELU_SEMICOARSENPFACTORY_KOKKOS_SHORT
+#include "MueLu_UseShortNames.hpp"
+
+  public:
+    //! @name Constructors/Destructors.
+    //@{
+
+    //! Constructor
+    SemiCoarsenPFactory_kokkos() : bTransferCoordinates_(false) { }
+
+    //! Destructor.
+    virtual ~SemiCoarsenPFactory_kokkos() { }
+    //@}
+
+    RCP<const ParameterList> GetValidParameterList() const;
+
+    //! Input
+    //@{
+
+    void DeclareInput(Level& fineLevel, Level& coarseLevel) const;
+
+    //@}
+
+    //! @name Build methods.
+    //@{
+
+    void Build (Level& fineLevel, Level& coarseLevel) const;
+    void BuildP(Level& fineLevel, Level& coarseLevel) const;
+
+    //@}
+
+    void BuildSemiCoarsenP(Level& coarseLevel, const LO NFRows, const LO NFNodes, const LO DofsPerNode,
+                           const LO NFLayers, const LO NCLayers, const LO LayerId[], const LO VertLineId[],
```

Comment:
Do not use raw pointer, you are asking for trouble, there is no reason not to have a Kokkos::View